scheduler/preemption.go

```diff
@@ -435,6 +439,96 @@ OUTER:
 	return filteredBestAllocs
 }
 
+// deviceGroupAllocs represents a group of allocs that share a device
+type deviceGroupAllocs struct {
+	allocs []*structs.Allocation
+
+	// deviceInstances tracks the number of instances used per alloc
+	deviceInstances map[string]int
+}
+
+func newAllocDeviceGroup() *deviceGroupAllocs {
+	return &deviceGroupAllocs{
+		deviceInstances: make(map[string]int),
+	}
+}
+
+// PreemptForDevice tries to find allocations to preempt to meet devices needed
+// This is called once per device request when assigning devices to the task
+func (p *Preemptor) PreemptForDevice(ask *structs.RequestedDevice, devAlloc *deviceAllocator) []*structs.Allocation {
+
+	// Group allocations by device, tracking the number of
+	// instances used in each device by alloc id
+	deviceToAllocs := make(map[structs.DeviceIdTuple]*deviceGroupAllocs)
+	for _, alloc := range p.currentAllocs {
+		for _, tr := range alloc.AllocatedResources.Tasks {
+			// Ignore allocs that don't use devices
+			if len(tr.Devices) == 0 {
+				continue
+			}
+
+			// Go through each assigned device group
+			for _, device := range tr.Devices {
+				// Look up the device instance from the device allocator
+				deviceIdTuple := *device.ID()
+				devInst := devAlloc.Devices[deviceIdTuple]
+
+				// devInst can be nil if the device is no longer healthy
+				if devInst == nil {
+					continue
+				}
+
+				// Ignore if the device doesn't match the ask
+				if !nodeDeviceMatches(p.ctx, devInst.Device, ask) {
+					continue
+				}
+
+				// Store both the alloc and the number of instances used
+				// in our tracking map
+				allocDeviceGrp := deviceToAllocs[deviceIdTuple]
+				if allocDeviceGrp == nil {
+					allocDeviceGrp = newAllocDeviceGroup()
+					deviceToAllocs[deviceIdTuple] = allocDeviceGrp
+				}
+				allocDeviceGrp.allocs = append(allocDeviceGrp.allocs, alloc)
+				allocDeviceGrp.deviceInstances[alloc.ID] += len(device.DeviceIDs)
+			}
+		}
+	}
+
+	neededCount := ask.Count
+
+	// Examine matching allocs by device
+	for deviceIDTuple, allocsGrp := range deviceToAllocs {
+		// First group and sort allocations using this device by priority
+		allocsByPriority := filterAndGroupPreemptibleAllocs(p.jobPriority, allocsGrp.allocs)
```

Comment:
A simple sort by priority on the outer loop also won't work because of the constraint that preemption has to be from the same device. You could end up with 
```
devicesToAllocs {
   matchingDev1: [(123, priority_1), (234, priority_70)],
   matchingDev2: [(345, priority_10), (456, priority_15)],
}
```
So we'd have to calculate the combination of preemptible allocs in each matching device, score them and then choose one. I made the tradeoff here of choosing a matching device at random, and then choosing the lowest priority allocations using that device, rather than calculating all combinations across all matching devices.

I thought this was a reasonable tradeoff to make as long as its documented. It is harder for operators to follow exactly why a particular device was chosen with this approach, so that's an argument for the exhaustive search method mentioned above.

Reply:
Oof... Makes sense. A potentially easy improvement that comes to mind is that instead of returning on line 523, you instead add the slice of preemptions to a `[][]*Allocations]` and do a continue on the outer for loop over device groups. This way at the end of the for loop we have the lowest priority sets of allocations per device group. Then we can go through each entry and come up with a score, and return the lowest cost one.